### PR TITLE
Amended to export a hof-behaviour.

### PR DESCRIPTION
**Pull request by joefitter**
Amended name to hof-behaviour-summary-page. Removed unnecessary deps, added hof-form-controller as devDep for tests

### CODE REVIEW

**Comment by joefitter** [comment]
@lennym how does this make you feel? Feel free to suggest a better name too!

File: package.json [hunk 1]
```diff
@@ -10,14 +10,14 @@
...
   "author": "",
   "license": "MIT",
   "bugs": {
-    "url": "https://github.com/UKHomeOfficeForms/hof-confirm-controller/issues"
+    "url": "https://github.com/UKHomeOfficeForms/hof-behaviour-summary-page/issues"
   },
-  "homepage": "https://github.com/UKHomeOfficeForms/hof-confirm-controller#readme",
+  "homepage": "https://github.com/UKHomeOfficeForms/hof-behaviour-summary-page#readme",
   "dependencies": {
     "hof-controllers": "^6.0.2",
```
Comment (lennym) [inline package.json]: There's a good chance we don't need this anymore.

**Comment by lennym** [comment]
Seems legit to me. I'd probably want to publish a patch update with a deprecation warning on the old package name too.

File: package.json (hunk 1 above)
Comment (joefitter) [reply package.json]: right you are, missed one

**Comment by JoeChapman** [comment]
Nice one, LGTM

**Comment by lennym** [comment]
I'm looking over this, and thinking about how to make the developer experience for behaviours a bit smoother. Currently we enforce this bit of code as boilerplate: `SuperClass => class extends SuperClass` and require the user to perform the mixing-in themselves in order to be able to test their behaviours.

Perhaps there's scope for a small `hof-behaviour` boilerplate module that will take a simple abstract class and turn it into a behaviour, as well as having a helper method that yields the behaviour applied to a form controller for testing.

So this example would look like:

```js
const Behaviour = require('hof-behaviour');
module.exports = Behaviour(class ConfirmBehaviour {
  locals(req, res) {...}
  ...
});
```

And the tests could do something like: 

```js
const Confirm = require('../');
const Behaviour = require('hof-behaviour');
const Controller = Behaviour.mix(Confirm);

// run tests on Controller class
```

That has the advantage of keeping the `mixwith` implementation opaque to the user, as well as not needing the additional dependencies in the top level of the tree.

Thoughts?

--

Edit: maybe rename `Behaviour.mix` to `Behaviour.test` to better indicate that it's intended for getting a test instance and not general use?

**Comment by JoeChapman** [comment]
@lennym  I think that is a good idea. 
I am also in favour of having a helper method exposed for implementation...
In tests
```js
const Controller = Behaviour.test(Confirm);
```
For general use
```js
const Controller = Behaviour.mix(Confirm);
```

**Comment by joefitter** [comment]
@lennym I like this approach, it was bugging me slightly that developers would have to use the mixin boilerplate.

```js
Behaviour.mix(class Confirm {...})

vs

Behaviour(class Confirm {...})
```
There isn't a lot of difference between these, but the benefit of the latter would be the developer wouldn't have to remember to use the method name, and just uses the default export - a la bootstrap and wizard, so I would be inclined to go with that.

Thoughts?

**Comment by lennym** [comment]
Any implementation for general use would need to a) define a base class to compose onto (this can possibly default to `hof-form-controller`, b) allow composition of multiple behaviours.

It also leads people down the road of defining their own controllers, which isn't necessarily something we want to encourage.

As such, I'd shy away from it for now. people can still use `mixwith` directly to do composition where needed, and any API we implemented would be practically identical to that anyway.

**Comment by lennym** [comment]
It's likely to be a 10 line module (if that), so I will knock together a PoC implementation once I'm done with what I'm doing now.

**Comment by JoeChapman** [comment]
@joefitter that is a good point
On a slightly different note, could there be a need for a user to apply behaviour to a behaviour class or only require a partial behaviour class?
Such as...
```js
Behaviour.mix(class Confirm {...}).and({
   customMethod() {},
   shadowMethod() {}
})
```
of 
```js
Behaviour.mix(class Confirm {...}).get({
   existingMethod() {}
})
```

**Comment by JoeChapman** [comment]
I am aware I may well be leagues behind current thinking on this.

**Comment by lennym** [comment]
I think you're possibly mudding the concerns a little there @JoeChapman. My proposals were purely around writing and testing behaviours, and not really trying to touch on how they're used at all.

That is currently fully implemented already by @joefitter in `hof-form-wizard` and at the moment I think we're generally trying to move away from implementations defining full controllers (this in fact being *the point* of behaviours), and so we should not worry about providing APIs to do this.

**Comment by lennym** [comment]
Unless - and on second reading I think you might be - you're talking about writing a behaviour that extends another behaviour?

**Comment by JoeChapman** [comment]
@lennym yes I am. I am aware that your original question was merely about hiding the implementation though. 
The reason I bring up behaviour extensions etc is that if we intend (in the future) to provide an api for these types of uses, then using methods such as `.test, .mix, .and, .get` to interact with Behaviours becomes a thing and as such, whether we use `Behaviour()` or `Behaviour.mx` is a moot point. 
However, for normal usage @joefitter has makes a good point, and for bog-standard use `Behaviour()` might well be the best way. It's all speculation

**Comment by lennym** [comment]
That's cool. I'm totally on board with trying to spec something that allows composition of behaviours in a way that *returns a behaviour*. I think we should shy away from any API (*except* `test`) that *returns a controller*.

Does that make sense?

**Comment by lennym** [comment]
With that in mind, how about:

```js
Behaviour(<class>) 
// returns class wrapped in behaviour boilerplate
Behaviour.extend(<class>, <...behaviours>) 
// returns class as a behaviour extending other behaviours
Behaviour.test(<behaviour>, [<controller>]) 
// returns a controller formed of a behaviour composed onto an optional base controller (defaulting to `hof-form-controller`)
```

**Comment by lennym** [comment]
I don't think what I've described above is possible, which is annoying.

I can't find a way to implement a function which turns a class into a mixin.

**Comment by joefitter** [comment]
I was thinking about it on my way back, and you might be right - the mixin needs to be a function that accepts a SuperClass, so the helper would need to take a class and return a function which extends said class from the SuperClass - problem being the class is already formed at that point so you can't extend it from anything

**Comment by joefitter** [comment]
Unless, we define a behaviour as a set of methods attached to a standard object, and use this to create the class in the helper... @lennym @JoeChapman thoughts?

**Comment by joefitter** [comment]
Ah bum - that wouldn't work as `super` would throw an error in a plain object :(

**Comment by joefitter** [comment]
Another thought - helper accepts class, returns function which accepts super class, this then creates a new class which extends from the super class, and some kind of dirty prototype method copying takes place from the class you pass in to the new class, the latter is returned. Not sure if this would work or exactly how to go about it without trying

**Comment by lennym** [comment]
I feel like once we're into that level of dirtiness then we might as well just deal with the single line of boilerplate and have things be nice.

**Comment by joefitter** [comment]
Agree. `super` isn't a thing unless you are actually extending in the class you use it in. Looks like this isn't going to be possible :(

**Comment by joefitter** [comment]
@lennym is this G2G as is in that case?

**Comment by lennym** [comment]
Yeah, I think so.

**Comment by joefitter** [comment]
👍

**Pull request merged**